src/clj/puppetlabs/puppetserver/ring/middleware/params.clj

```diff
@@ -0,0 +1,93 @@
+(ns puppetlabs.puppetserver.ring.middleware.params
+  (:require [ring.util.codec :as codec]
+            [ring.util.request :as req]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; COPY OF RELEVANT FUNCTIONS FROM UPSTREAM ring.middleware.params LIBRARY
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; This namespace is basically just here to provide an implementation of
+;; the `params-request` middleware function that supports a String representation
+;; of the body of a request.  The upstream library requires the body to be of
+;; a type that is compatible with Clojure's IOFactory, which forces us to read
+;; the request body into memory twice for requests that we have to pass down
+;; into the JRuby layer.  (Technically, the Ring specification states that the
+;; body must be an InputStream, so the maintainer of the upstream library was
+;; reluctant to accept any sort of upstream PR to work around this issue.)
+;;
+;; All of this code is copied from the upstream library, and there is just
+;; one very slight modification (see comment in `assoc-form-params` function)
+;; that allows us to avoid reading the body into memory twice.
+;;
+;; In the future, if we can handle the query parameter parsing strictly on the
+;; Clojure side and remove that code from the Ruby side, we should be able to
+;; get rid of this.  That will be much easier to consider doing once we're able
+;; to get rid of the Rack/Webrick support.
+;;
+;; If that happens, we should delete this namespace :)
+;;
+
+(defn parse-params [params encoding]
+  (let [params (codec/form-decode params encoding)]
+    (if (map? params) params {})))
+
+(defn content-type
+  "Return the content-type of the request, or nil if no content-type is set."
+  [request]
+  ;; NOTE: in the latest version of ring-core, they only look in
+  ;;  the headers map for the content type.  They no longer fall
+  ;;  back to looking for it in the main request map.
+  (if-let [type (or (get-in request [:headers "content-type"])
```

Comment:
Not sure where all we'd be broken by not compensating for this elsewhere in Puppet Server but maybe it would be better for us to just "deal with it" elsewhere in our code?  I actually like what Ring did in eliminating the ambiguity around having this value represented in two different places in the map - which one should you honor if they happen to be different, etc?  One place I know this would affect is the hack that we did for the "file_bucket_file" binary content support.  To be compatible there, we might have to move the hack down from the `compojure/routes" into the master request handler, for example.